src/staticdata.c

```diff
@@ -1741,8 +1741,20 @@ static void jl_write_values(jl_serializer_state *s) JL_GC_DISABLED
                 else {
                     newm->nroots_sysimg = m->roots ? jl_array_len(m->roots) : 0;
                 }
-                if (m->ccallable)
-                    arraylist_push(&s->ccallable_list, (void*)reloc_offset);
+                if (m->ccallable) {
+                    int should_push = !jl_options.trim;
+                    for (size_t i = 0; !should_push && i < jl_entrypoint_mis->len  ; i++) {
+                        jl_method_instance_t* mi = (jl_method_instance_t*)jl_entrypoint_mis->items[i];
+                        assert(jl_is_method_instance(mi));
+                        jl_method_t *ccallable = mi->def.method;
+                        if (ccallable == m) {
```

Comment:
There's a corner case here where the MI != the `@ccallable` signature, which would still lead to a crash I think.

Can we check for that?

Reply:
I think we already assert that. 
https://github.com/JuliaLang/julia/blob/fb02e641066fe423c34ee8e4fec24b965662ce85/src/precompile_utils.c#L370-L394

Reply:
I can add a comment that mentions that though

Reply:
I don't see where that's asserted there - Can you spell it out for me?

Reply:
```julia
static void *jl_precompile_trimmed(size_t world)
{
    // array of MethodInstances and ccallable aliases to include in the output
    jl_array_t *m = jl_alloc_vec_any(0);
    jl_value_t *ccallable = NULL;
    JL_GC_PUSH2(&m, &ccallable);
    jl_method_instance_t *mi;
    for (size_t i = 0; i < jl_entrypoint_mis->len ; i++) {
        mi = (jl_method_instance_t*)jl_entrypoint_mis->items[i];
        if (mi == NULL)
            break;
        assert(jl_is_method_instance(mi));

        jl_array_ptr_1d_push(m, (jl_value_t*)mi);
        ccallable = (jl_value_t *)mi->def.method->ccallable; // If the method passed into jl_entrypoint has ccallable we always compile it. 
        if (ccallable)
            jl_array_ptr_1d_push(m, ccallable);
    }
    jl_cgparams_t params = jl_default_cgparams;
    params.trim = jl_options.trim;
    void *native_code = jl_create_native(m, NULL, &params, 0, /* imaging */ 1, 0,
                                         world, NULL);
    JL_GC_POP();
    return native_code;
}
```

Reply:
It means if we are compiling the entrypoint and if the method of that entrypoint has a ccallable defined on it, we will add both the entrypoint MI and the ccallable signature (it's an svec). We can maybe compare ccallables instead of methods.

Reply:
That's fine, but we shouldn't be imprecise about checking the entrypoint just because we happen to enqueue the ccallables in that way right now